### PR TITLE
Add challenge/response auth mechanism to the API server protocol

### LINKED ISSUE #91: Write support for the Zino 1 legacy user database and authentication mechanisms

### PR DESCRIPTION
**Pull request by lunkwill42**
Closes #91 .

This does not add all the functionality of the original, as it has the option to configure aspects of authentication through a config file.  We have not yet added a config file feature, so we have added this update to the task list in #63

### CODE REVIEW

File: src/zino/api/legacy.py [hunk 1]
```diff
@@ -22,12 +25,15 @@ def requires_authentication(func: Callable) -> Callable:
 class Zino1BaseServerProtocol(asyncio.Protocol):
     """Base implementation of the Zino 1 protocol, with a basic command dispatcher for subclasses to utilize."""
 
-    def __init__(self):
+    def __init__(self, secrets_file: Optional[Union[Path, str]] = "secrets"):
```
Comment (hmpf) [inline src/zino/api/legacy.py:28]: When there is a config-file the location of the "user database" needs to be configurable.

File: tests/api/auth_test.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        )
+
+    def test_when_challenge_response_is_bad_should_raise_error(self, secrets_file):
+        with pytest.raises(AuthenticationFailure):
+            assert authenticate(
+                "user1",
+                response="b3e9ba156949152c5a6f3334df86333b1daabfe9",
+                challenge="a97fbebb3eef2dfcf167645229c0c5fa9e92e3da",
+                secrets_file=secrets_file,
+            )
```
Comment (hmpf) [inline tests/api/auth_test.py:22]: These two look identical to me, why should one fail and the other succeed?

**Review by hmpf — COMMENTED** [review]
Has a test been copy-pasted with insufficient changes?

File: tests/api/auth_test.py (hunk 2 above)
Comment (lunkwill42) [reply tests/api/auth_test.py:22]: Ah, well, `b3e9ba156949152c5c6f3334cf86333b1daabfe9` != `b3e9ba156949152c5a6f3334df86333b1daabfe9`.  Do we need them to be obviously different to the human eye? :)

Comment (lunkwill42) [reply tests/api/auth_test.py]: alright, new version pushed

File: tests/api/auth_test.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                response="doesntmatter",
+                secrets_file=secrets_file,
+            )
+
+
+class TestGetChallenge:
+    def test_should_never_return_same_challenge(self):
+        count = 10
+        challenges = set(get_challenge() for _ in range(count))
+        assert len(challenges) == 10, "same challenge was produced more than once"
```
Comment (stveit) [inline tests/api/auth_test.py:49]: This is a weird test imo.

While its unlikely to generate the same value, the `secrets` library does not actually promise to not generate the same value twice. Promising unique values would technically make tokens less secure (or more predictable), as this would mean one possible number is removed every time its run.

The `get_challenge` wrapping function also does nothing to avoid duplicates. So this test technically tests for something that is not actually expected behavior.

Also, this test is technically "flaky", since there is a rare possibility that you actually get duplicates

File: tests/api/auth_test.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    def test_should_return_at_least_40_characters(self):
+        assert len(get_challenge()) >= 40
+
+
+def test_read_users_when_given_a_valid_file_then_a_proper_dict_should_be_returned(secrets_file):
+    users = read_users(secrets_file)
+    for user in ("user1", "user2", "user3"):
+        assert user in users
+        assert len(users[user]) == 40
```
Comment (stveit) [inline tests/api/auth_test.py:59]: In one test you check that a challenge is at least 40, and in the next you check that its exactly 40. This is a bit inconsistent, even though it works in this case since you know what the `secret_file` contains. Would be nice to be consistent between these tho for "tests as documentation" purposes, since this kinda implies a requirement for being a "proper" dict is that it contains secrets of exactly length 40.

Feel free to disagree. If this is supposed to test that the secrets in the `secret_file` are returned properly  from `read_users`, then checking the exact length makes some sense. But in that case it might make more sense to check the exact values instead.

The problem here might more be the naming of the test and not the actually testing tho

File: tests/api/conftest.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+@pytest.fixture
+def secrets_file(tmp_path):
+    name = tmp_path.joinpath("secrets")
+    with open(name, "w") as conf:
+        conf.write(
+            """user1 3c55d3cdc19876dfc8c0bb49da8c927a0ddff26d
+            user2 eb6fb35f5fbba6a6e43d3c893ad7a77dc793ceba
+
+            user3 c8a0b250edb2eabd9616b7b05a46e0ad28226fc2
+            """
```
Comment (stveit) [inline tests/api/conftest.py:13]: is the newline significant?

**Review by stveit — CHANGES_REQUESTED** [review]

File: tests/api/conftest.py (hunk 5 above)
Comment (lunkwill42) [reply tests/api/conftest.py:13]: No. It's there to ensure that random newlines in the input does not hold any significance to the parser.  This could, of course, be a separate test to make it more explicit.

File: tests/api/auth_test.py (hunk 4 above)
Comment (lunkwill42) [reply tests/api/auth_test.py:59]: > In one test you check that a challenge is at least 40, and in the next you check that its exactly 40. 

No. In one test, I verify that a *challenge issued by Zino* is at least 40 characters. In the next test, I verify that the *secrets* (no challenges are involved here) parsed from the test data are exactly as long as they were in the test data.

More explicitly, they could test that the secrets are identical to the inputs, as this test would not detect that the code changed the secrets into all `A` characters, for example :grin:

Comment (stveit) [reply tests/api/auth_test.py:59]: Ah, I see. I've misunderstood the connection between challenge and secrets

File: tests/api/conftest.py (hunk 5 above)
Comment (stveit) [reply tests/api/conftest.py:13]: yeah i think that intention should be made clearer

File: tests/api/auth_test.py (hunk 3 above)
Comment (lunkwill42) [reply tests/api/auth_test.py:49]: This test is mostly meant to safeguard against a n00b developer b0rking the `get_challenge()` implementation 10 years down the line into something that returns the same static value every time it is called.  It is not meant to be a spec that says that challenges should *NEVER* be the same.

I will personally buy you a very expensive craft beer if you ever observe the secrets library to generate the same value twice in a row during this test :)

Comment (stveit) [reply tests/api/auth_test.py:49]: > It is not meant to be a spec that says that challenges should _NEVER_ be the same.

then you should rename the test to not be `test_should_never_return_same_challenge`, Theres not really any other way to interpret "never return same challenge"

**Review by stveit — COMMENTED** [review]
Also the code smells should be resolved

**Comment by lunkwill42** [comment]
> Also the code smells should be resolved

:-1: The code smells are triggered in files not touched by this PR.  SonarCloud is apparently on drugs...

File: tests/api/auth_test.py (hunk 3 above)
Comment (lunkwill42) [reply tests/api/auth_test.py]: done

**Comment by stveit** [comment]
> > Also the code smells should be resolved
> 
> 👎 The code smells are triggered in files not touched by this PR. SonarCloud is apparently on drugs...

lmao

**Review by stveit — APPROVED** [review]
